### ISSUE TITLE
specifying serializers for API::V2

### ISSUE THREAD

**Issue by dbwieland18** [fotinakis/jsonapi-serializers#18]
I'm using Active Model Serializers for V1 of my app, while I'm wanting to implement jsonapi-serializers for V2. Having trouble getting things routed correctly since both gems want serializers defined in a 'serializers' directory. Is there a way around this or is it dependent on overriding serializer discovery, as mentioned in 'unfinished business'?

**Comment by fotinakis** [comment]
There's not a great way to do this currently, no, but I'd like to support it.

Currently, jsonapi-serializers discovers serializers by just doing `"#{object.class.name}Serializer"` — so, if your models are also in a versioned namespace, it'll just work. However, versioning models is not common, so we definitely need a way to override serializer discovery. I guess you could pretty easily monkey-patch the `self.find_serializer_class_name` method right now, but that's of course ugly.

Seems like this has to be something you'd pass in to the `serialize` method every time, since the serializer class can't be inferred from the objects that are being serialized. So something like: `JSONAPI::Serializer.serialize(object, finder: V2SerializerFinder)` or something. Does that feel right?

**Comment by l8nite** [comment]
@fotinakis yea that feels right.

I did something similar in my `JsonApiNet` library for deserialization with a custom 'type resolver' - 

Basically, you give it any object that implements the "ResolveType" method, and it returns the class. 

From the [README](https://github.com/l8nite/JsonApiNet/blob/master/README.md)

```
public class BarneyTypeResolver : IJsonApiTypeResolver {
  public Type ResolveType(string typeName) {
    if (typeName == "rain_drops") {
      return typeof(LemonDrop);
    }

    return null;
  }
}
Then you can use it like so:

var lemonDrop = JsonApi.ResourceFromDocument<LemonDrop>(json, new BarneyTypeResolver())
```

**Comment by l8nite** [comment]
In fact, if you don't mind the slight violation of concerns (why should an object know anything about how it is being serialized?), you could just define an interface that the object would implement:

```
def self.find_serializer_class_name(object)
  if object.respond_to?(:jsonapi_serializer_class_name)
    return object.jsonapi_serializer_class_name
  end
end
```

**Comment by fotinakis** [comment]
I had thought of that last one but I don't think it works because, for
multiple versions, you need to be able to support serializing the same
object with different serializers dynamically, it can't just be a single
serializer return.

**Comment by fotinakis** [comment]
Quick update: I'm working on support for explicit serializer discovery in PR https://github.com/fotinakis/jsonapi-serializers/pull/37 — this doesn't totally solve this issue, because there's no state passed in to know whether to serialize v1 or v2, but it gets us closer.

**Comment by smackesey** [comment]
I'm running into similar problems. @fotinakis, your suggestion of passing in a "finder" instance to `serialize` seems like a good solution to the problem. As I imagine this, a finder can be any object that responds to `call`, takes a single argument (the object to be serialized), and returns a serializer class.

Implementing this shouldn't be too difficult, but the default resolution mechanism should probably be wrapped as a Finder object as well. If this approach sounds OK to you, I will submit a PR for it.

**Comment by fotinakis** [comment]
Cool, I would definitely be willing to review a PR for that! Since we don't have an explicit style-guide yet, please try to follow the local code style which should be pretty consistent at this point.

**Comment by fotinakis** [comment]
Fixed in #79
